### ISSUE TITLE
mAP drops down when training (2 classes)

### ISSUE THREAD

**Issue by asverlov** [AlexeyAB/darknet#5033]
I'm new to this, and maybe missing something very basic, but why would mAP go down like that?

![image](https://user-images.githubusercontent.com/2652179/76710332-14bda580-66dd-11ea-9daa-3a8466761c47.png)
I've taken two image collections: egohands and whideface, used in other darknet projects, randomly tested them with yolo_mark to make sure that labels are ok  (they are) and tried to do some training on them with openimage weights. 

Training image sizes are around 1280x720 -1000x600 mostly, some go down to 800 x 400.

Config:
```
[net]
# Testing
# batch=1
# subdivisions=1
# Training
batch=64
subdivisions=16
width=512
height=512
```
Two class names are referenced through data file.  In config classes are set to 2, filters to 21 in the proper three places.
Train sets are big and each only has it's own boxes, validation set is tiny and has both classes marked. 
I've tried to start training based on opeinimage weights ,  on darknet53.conv.74 and on nothing at all.
And in some cases I end up with faces and in other with hands
![image](https://user-images.githubusercontent.com/2652179/76713230-8bff3380-66f5-11ea-8d05-7db0fcd823e2.png)

![image](https://user-images.githubusercontent.com/2652179/76713290-f0ba8e00-66f5-11ea-939b-91ab1fb3c93e.png)

Should I keep training, increase the validation set (if it affects anything), change size in config (I've tried 416x416) or try something else?

Any ideas are appreciated!

**Comment by lsd1994** [comment]
Are the annotations right? You should label both classes in train&valid dataset.

**Comment by asverlov** [comment]
> Are the annotations right? You should label both classes in train&valid dataset.
One training dataset (egohands) has only hands labeled and another (widefaces) only faces. When used by themselves they work great. I don't think it's feasible to label both classes in each training dataset - they are huge. 
Validation data has both classes labeled.

**Comment by asverlov** [comment]
So the question is: should every instance of the class we are training for be labeled in every image.  i.e. dataset with faces has hands in some images and they are not labeled.  Is it something which might cause a problem?

**Comment by lsd1994** [comment]
Yes, you should label all faces and hands in one image.

**Comment by asverlov** [comment]
Then it sounds like a systematic problem.  
We have one dataset of 10,000 correctly labeled face (and only face) images and another dataset of 10,000 correctly labeled hand (and only hand) images -  that's how you get them from the public domain. And by the nature of the beast  every image can potentially have other classes there, especially when they belong to a common "person" parent. 
Theoretically there is should enough data to build a network here if everything is correctly labeled.
So the question is: how to tell the training algorithm where's what. Going manually over 20,000 images and 'quickly' relabeling them doesn't' sound like a working proposition.... and a lot of counterproductive human labor. 
Maybe @AlexeyAB  can chime in  and shed some light on the inner workings of the system.

**Comment by AlexeyAB** [comment]
* Or you should label all objects in all images
* Or you should change the source code - add feature for disabling training for the specified classes on specified images

**Comment by asverlov** [comment]
I was also thinking about training on only one class at a time, create weights1 and weights2 and either use them in 
- YOLT way - just detect twice in the final app (loose time) or
- use weight1 to add labels to dataset2 (and other way around) and then retrain for single weight file (loose precision?) or
- adding a flag for non-strict training, which seems to be the best.  It could dynamically shrink class list to only those specified in .txt label files, if such thing is possible.   Can you suggest what source files to start looking at? Thanks!
